plover/steno.py

```diff
@@ -21,6 +21,78 @@
 _NUMBERS = set('0123456789')
 _IMPLICIT_NUMBER_RX = re.compile('(^|[1-4])([6-9])')
 
+def compare_normalize_stroke(x, y):
+    for i in range(len(x.strokes)):
+        if i > len(y.strokes) - 1:
+            return 1
+        x_norm = normalize_stroke(x.strokes[i])
+        y_norm = normalize_stroke(y.strokes[i])
+        if less_than(x_norm, y_norm):
+            return -1
+        elif less_than(y_norm, x_norm):
+            return 1
+    if len(x.strokes) < len(y.strokes):
+        return -1
+    return 0
+
+def less_than(x, y):
+    x_right_side = True
+    y_right_side = True
+    for i in range(len(x)):
```

Comment:
Okay -- more logical now, since you have multiple levels of comparison here it may as well be represented.

I'd say name your two functions "cmp_normalized_steno" and "cmp_normalized_strokes"

This is untested (might not even run O.O)

```python
def cmp_normalized_steno(x, y):
    for stroke_x, stroke_y in zip(x, y):
        cmp_strokes = cmp_normalized_strokes(x, y)
        if cmp_strokes:
            return cmp_strokes
    return cmp(len(x.strokes), len(y.strokes))

def cmp_normalized_strokes(x, y):
    # Turn string into a 'stack'
    x = list(reverse(x))
    y = list(reverse(y))

    # Returns key index and whether it is on the left of the board.
    def get_key_index(key, side):
        if side == 1:  # Right side, easiest case.
            return system.KEY_ORDER.get('-%s' % key), 1
        elif key == '-'  # Center delimiter
            return None, 1
        elif key in system.IMPLICIT_HYPHENS:  # Detect center keys
            index = (
                system.KEY_ORDER.get(key) or
                system.KEY_ORDER.get('%s-' % key) or
                system.KEY_ORDER.get('-%s' % key)
            )
            assert index is not None, 'Invalid implicit key found: %s' % key
            return index, 0
        elif side == 0:
            # Case where we last found a center key, but now can't.
            # It implies our first right side key.
            return system.KEY_ORDER.get('-%s' % key), 1
        else: # Must be a left-side key.
            return system.KEY_ORDER.get('%s-' % key), -1
    
    # -1: left, 0: center, 1: right
    x_side = -1
    y_side = -1

    # While x and y both have content...
    while x and y:
        x_index, x_side = get_key_index(x.pop(), x_side)
        if x_index is None:
            x_index, x_side = get_key_index(x.pop(), x_side)
        y_index, y_side = get_key_index(x.pop(), y_side)
        if y_index is None:
            y_index, y_side = get_key_index(x.pop(), y_side)
        index_compare = cmp(x_index, y_index)
        if index_compare: # If not equal, we have a result.
            return index_compare
    # One will be zero at least.
    return cmp(len(x), len(y))
```